python/ray/data/_internal/logical/optimizers.py

```diff
@@ -19,30 +30,211 @@ class PhysicalOptimizer(Optimizer):
     (3) physical optimization: optimize physical operators.
     """
-    optimized_logical_dag = LogicalOptimizer().optimize(logical_dag)
-    physical_dag = Planner().plan(optimized_logical_dag)
-    return PhysicalOptimizer().optimize(physical_dag)
+    logical_plan = LogicalOptimizer().optimize(logical_plan)
+    physical_plan = Planner().plan(logical_plan)
+    return PhysicalOptimizer().optimize(physical_plan)
+
+
+class OperatorFusionRule(Rule):
+    """Fuses linear chains of compatible physical operators."""
+
+    def apply(self, plan: PhysicalPlan) -> PhysicalPlan:
+        self._op_map = plan.op_map.copy()
+        # Do DFS fusion.
+        root = self._apply(plan.dag)
+        return PhysicalPlan(root, self._op_map)
+
+    def _apply(self, op: PhysicalOperator) -> PhysicalOperator:
+        """Performs DFS fusion of linear chains of physical map operators, provided that
+        they are pairwise-compatible.
+
+        Args:
+            op: The op that we're trying to fuse with its input.
+        """
+        upstream_ops = op.input_dependencies
+        # Fuse with upstream ops while possible.
+        while len(upstream_ops) == 1 and self._can_fuse(op, upstream_ops[0]):
+            # Fuse operator with its upstream op.
+            op = self._fuse(op, upstream_ops[0])
+            upstream_ops = op.input_dependencies
+        # Can no longer fuse with upstream ops, proceed up the DAG.
+        op._input_dependencies = [
+            self._apply(upstream_op) for upstream_op in upstream_ops
+        ]
+        return op
+
+    def _can_fuse(self, down_op: PhysicalOperator, up_op: PhysicalOperator) -> bool:
+        """Returns whether the provided downstream operator can be fused with the given
+        upstream operator.
+
+        We currently support fusing two operators if the following are all true:
+            * They are both MapOperators.
+            * They either use the same compute configuration, or the upstream operator
+              uses a task pool while the downstream operator uses an actor pool.
+            * If both operators involve callable classes, the callable classes are
+              the same class AND constructor args are the same for both.
+            * They have compatible remote arguments.
+        """
+        from ray.data._internal.execution.operators.map_operator import MapOperator
+        from ray.data._internal.logical.operators.map_operator import AbstractMap
+
+        # We only support fusing MapOperators.
+        if not isinstance(down_op, MapOperator) or not isinstance(up_op, MapOperator):
+            return False
+
+        down_logical_op = self._op_map[down_op]
+        up_logical_op = self._op_map[up_op]
+
+        # Allow fusing tasks->actors if the resources are compatible (read->map), but
+        # not the other way around. The latter (downstream op) will be used as the
+        # compute if fused.
+        if (
+            is_task_compute(down_logical_op._compute)
```

Comment:
shouldn't we check `isinstance(down_logical_op, AbstractMap)` before `is_task_compute(down_logical_op._compute)` ?

Reply:
Hmm I think that's a good defensive check to add. We already have an `isinstance(down_op, MapOperator)` check above, which limits the logical op to `Read()` or `AbstractMap()`, and you currently can't have a `Read()` as the downstream op (it's always a source op), but that's a lot of assumptions. And there's a good chance that we'll introduce more logical operations that will result in a `MapOperator` physical operator, so that's a good defensive check for guarding against future failures.

Actually, how about we add an upfront logical op check like this right after fetching the logical ops?
```python
if not isinstance(down_logical_op, AbstractMap) or not isinstance(up_logical_op, (Read, AbstractMap)):
   return False
```
Then we can assume that constraint for the rest of the function and tweak that condition as we add more logical ops that result in physical `MapOperator`s.

Reply:
yeah SGTM.